### PR TITLE
Add a new feature flag which could switch the virtual delete_vm function when instance uprade

### PR DESCRIPTION
**Pull request by jianqiu**
When instance upgrade, bosh director will first delete vm from infrastructure then recreate the vm. Under some special conditions, for example, in SoftLayer, we don`t want to real delete vm from Softlayer, bosh softlayer cpi would leverage the os_reload provided by Softlayer to upgrade the stemcell/instance vms.
So, I just submit this PR which including a new feature flag to enable virtual delete vm when instance upgrade.

### CODE REVIEW

**Comment by jianqiu** [comment]
@cppforlife @mattcui @maxim
A new PR to trace this enhancement.

**Comment by jianqiu** [comment]
@cppforlife @maximilien
Could you please look into this PR? Thanks!

File: bosh-director/lib/bosh/director/vm_deleter.rb [hunk 1]
```diff
@@ -26,7 +27,11 @@ def delete_for_instance(instance)
 
     def delete_vm(vm_cid)
       @logger.info('Deleting VM')
-      @error_ignorer.with_force_check { @cloud.delete_vm(vm_cid) }
+      @error_ignorer.with_force_check {
```
Comment (cppforlife) [inline bosh-director/lib/bosh/director/vm_deleter.rb]: - use `do-end` syntax for multiple lines
- also add a test

File: bosh-director/lib/bosh/director/vm_creator.rb [hunk 2]
```diff
@@ -156,7 +156,9 @@ def create(instance_model, stemcell, cloud_properties, network_settings, disks,
       @logger.error("error creating vm: #{e.message}")
       if vm_cid
         parent_id = add_event(instance_model.deployment.name, instance_model.name, 'delete', vm_cid)
-        @vm_deleter.delete_vm(vm_cid)
+        unless Config.enable_virtual_delete_vms
```
Comment (cppforlife) [inline bosh-director/lib/bosh/director/vm_creator.rb]: - add test

**Comment by medvedzver** [comment]
Hi!
Could you please make a pull request as a single commit?
Thanks!

**Comment by cppforlife** [comment]
To clarify, please git squash the commits into a single commit so it's easy to merge. We had difficulties merging this in as is.

File: bosh-director/lib/bosh/director/vm_deleter.rb (hunk 1 above)
Comment (cppforlife) [inline bosh-director/lib/bosh/director/vm_deleter.rb]: to clarify unit test should be sufficient here. see vm_creator_spec.rb example.

**Comment by jianqiu** [comment]
@cppforlife @mattcui @medvedzver  @maximilien
It is pretty hard to squash all commits into single commit, so just closing it and submitted a new clean PR, sorry for the inconvenience caused for you.
https://github.com/cloudfoundry/bosh/pull/1251

**Comment by voelzmo** [comment]
@medvedzver @cppforlife For the next time: If you prefer squashed PRs, you can now configure github to do this for you: https://help.github.com/articles/about-pull-request-merge-squashing/ 
You might want to enable this in the settings.

**Pull request closed without merging**